### PR TITLE
SerialCommHub - a serial port concurrent access multiplexer

### PR DESCRIPTION
**Pull request by corneliusclaussen**
Multiplexes access from multiple modules to the same serial port for ModBus and ModBus-like raw requests.

Signed-off-by: LAD101work <96466764+LAD101work@users.noreply.github.com>
Signed-off-by: Cornelius Claussen <cc@pionix.de>

### CODE REVIEW

File: modules/CMakeLists.txt [hunk 1]
```diff
@@ -25,6 +25,7 @@ ev_add_modules(
     PersistentStore
     PN532TokenProvider
     Setup
+    SerialCommHub
```
Comment (a-w50) [inline modules/CMakeLists.txt:28]: Concerning the naming, is this really a universal serial communication hub? Looking at the at the manifest, it looks more like a ModbusMultiplexer

File: modules/SerialCommHub/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+#   - ${MODULE_NAME}: module name
+ev_setup_cpp_module()
+
+# ev@bcc62523-e22b-41d7-ba2f-825b493a3c97:v1
+# insert your custom targets and additional config variables here
+
+target_link_libraries(${MODULE_NAME}
+    PRIVATE
+    everest::modbus
+    date::date
```
Comment (a-w50) [inline modules/SerialCommHub/CMakeLists.txt:16]: Is date needed here?

**Review by hikinggrass — APPROVED** [review]

File: modules/SerialCommHub/main/serial_communication_hubImpl.cpp [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+    case 1:
+        cfg.set_parity(everest::connection::SerialDeviceConfiguration::Parity::Odd);
+        break;
+    case 2:
+        cfg.set_parity(everest::connection::SerialDeviceConfiguration::Parity::Even);
+        break;
+    }
+
+    serial_device = std::unique_ptr<everest::connection::SerialDevice>(new everest::connection::SerialDevice(cfg));
+    modbus_connection =
```
Comment (a-w50) [inline modules/SerialCommHub/main/serial_communication_hubImpl.cpp:40]: same here, only pass *serial_device
RTUConnection actually owns a reference to the serial device - therefor it is also importent, that the unique_ptr modbus_connection goes out of scope before serial_device (order of member variable listing is important!)

**Review by a-w50 — COMMENTED** [review]
See comments

File: modules/CMakeLists.txt (hunk 1 above)
Comment (corneliusclaussen) [reply modules/CMakeLists.txt:28]: Currently it supports modbus and raw access for similar protocols that are not really modbus. full raw could be added as well

File: modules/SerialCommHub/CMakeLists.txt (hunk 2 above)
Comment (corneliusclaussen) [reply modules/SerialCommHub/CMakeLists.txt:16]: probably that's a left over, will remove

File: modules/SerialCommHub/main/serial_communication_hubImpl.hpp [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    std::unique_ptr<everest::connection::SerialDevice> serial_device;
+    std::unique_ptr<everest::connection::RTUConnection> modbus_connection;
+    std::unique_ptr<everest::modbus::ModbusRTUClient> modbus_client;
+    std::mutex serial_mutex;
+
+    everest::connection::SerialDeviceConfiguration::BaudRate baudrate_from_int(int baudrate);
+    template <typename T, typename U> void append_array(std::vector<T>& m, const std::vector<U>& a);
+    bool check_crc(const uint8_t* buffer, size_t len);
+    std::vector<uint16_t> toLittleEndian16(const std::vector<uint8_t> response);
+    std::vector<int> toLittleEndianInt(const std::vector<uint8_t> response);
```
[the remark this answers is not in the document]
Comment (corneliusclaussen) [reply modules/SerialCommHub/main/serial_communication_hubImpl.hpp:83]: const refs: as far as i could see this is in autogenerated code

File: modules/SerialCommHub/main/serial_communication_hubImpl.cpp [hunk 5]
```diff
@@ -0,0 +1,367 @@
...
+        break;
+    case 1:
+        cfg.set_parity(everest::connection::SerialDeviceConfiguration::Parity::Odd);
+        break;
+    case 2:
+        cfg.set_parity(everest::connection::SerialDeviceConfiguration::Parity::Even);
+        break;
+    }
+
+    serial_device = std::unique_ptr<everest::connection::SerialDevice>(new everest::connection::SerialDevice(cfg));
```
[the remark this answers is not in the document]
Comment (corneliusclaussen) [reply modules/SerialCommHub/main/serial_communication_hubImpl.cpp:39]: how would that line look like?

File: modules/SerialCommHub/main/serial_communication_hubImpl.cpp (hunk 3 above)
Comment (corneliusclaussen) [reply modules/SerialCommHub/main/serial_communication_hubImpl.cpp:40]: yes, that would all be easier if the modbus and serial library would support a constructor without arguments as well

File: modules/SerialCommHub/main/serial_communication_hubImpl.cpp [hunk 6]
```diff
@@ -0,0 +1,367 @@
...
+    modbus_connection =
+        std::unique_ptr<everest::connection::RTUConnection>(new everest::connection::RTUConnection(*serial_device));
+    modbus_client =
+        std::unique_ptr<everest::modbus::ModbusRTUClient>(new everest::modbus::ModbusRTUClient(*modbus_connection));
+
+    try {
+        serial_device->open();
+    } catch (const everest::connection::exceptions::tty::tty_error& e) {
+        EVLOG_AND_THROW(Everest::EverestConfigError(fmt::format("Cannot open serial port {}.", config.serial_port)));
+    }
```
[the remark this answers is not in the document]
Comment (corneliusclaussen) [reply modules/SerialCommHub/main/serial_communication_hubImpl.cpp:49]: fixed

File: modules/SerialCommHub/main/serial_communication_hubImpl.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (corneliusclaussen) [reply modules/SerialCommHub/main/serial_communication_hubImpl.hpp:83]: rest is fixed

File: modules/SerialCommHub/main/serial_communication_hubImpl.cpp [hunk 7]
```diff
@@ -0,0 +1,367 @@
...
+}
+
+void serial_communication_hubImpl::ready() {
+}
+
+// Commands
+
+types::serial_comm_hub_requests::StatusCodeEnum
```
[the remark this answers is not in the document]
Comment (corneliusclaussen) [reply modules/SerialCommHub/main/serial_communication_hubImpl.cpp:57]: We could, but I don't like renaming too much. It makes it harder to understand that we are using the everest external types for the interface here

File: modules/SerialCommHub/main/serial_communication_hubImpl.cpp [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+        const uint8_t crc_high = (uint8_t)(crc_sum & 0x00FF);
+
+        message.push_back(crc_low);
+        message.push_back(crc_high);
+    }
+
+    int bytes_written = 0;
+    {
+        // write to device
+        std::scoped_lock lock(serial_mutex);
```
[the remark this answers is not in the document]
Comment (corneliusclaussen) [reply modules/SerialCommHub/main/serial_communication_hubImpl.cpp:80]: yes, but std::lock_guard<std::mutex> lock(serial_mutex); has a quite ugly syntax, while scoped_lock is quite readable and clear that it actually is scoped

File: modules/SerialCommHub/main/serial_communication_hubImpl.cpp [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+
+    // do we need to add CRC at the end?
+    if (add_crc16) {
+        const uint16_t crc_sum = everest::modbus::utils::calcCRC_16_ANSI(message.data(), message.size());
+        const uint8_t crc_low = (uint8_t)(crc_sum >> 8);
+        const uint8_t crc_high = (uint8_t)(crc_sum & 0x00FF);
+
+        message.push_back(crc_low);
+        message.push_back(crc_high);
+    }
```
[the remark this answers is not in the document]
Comment (corneliusclaussen) [reply modules/SerialCommHub/main/serial_communication_hubImpl.cpp:113]: fixed

File: modules/SerialCommHub/main/serial_communication_hubImpl.cpp [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+    uint8_t hi, lo;
+    for (auto it = response.begin(); it != response.end();) {
+        hi = *it++;
+        if (it == response.end())
+            break;
+        lo = *it++;
+        r.push_back((uint16_t)((hi << 8) | lo));
+    }
+    return r;
+}
```
[the remark this answers is not in the document]
Comment (corneliusclaussen) [reply modules/SerialCommHub/main/serial_communication_hubImpl.cpp:308]: comments added

File: modules/SerialCommHub/main/serial_communication_hubImpl.cpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (corneliusclaussen) [reply modules/SerialCommHub/main/serial_communication_hubImpl.cpp:39]: ah github removes the sharp brackets

**Pull request merged**